fedimint-cli/src/lib.rs

```diff
@@ -516,12 +532,34 @@ impl FedimintCli {
                 hash: env!("FEDIMINT_BUILD_CODE_VERSION").to_string(),
             }),
             Command::Client(ClientCmd::Restore { secret }) => {
-                let client = cli
+                let mut client_builder = cli
                     .build_client_builder(&self.module_inits, None)
                     .await
-                    .map_err_cli_msg(CliErrorKind::GeneralFailure, "failure")?
-                    .build_restoring_from_backup(ClientSecret::<PlainRootSecretStrategy>::new(
-                        secret,
+                    .map_err_cli_general()?;
+                let client_secret = match client_builder
+                    .load_decodable_client_secret::<[u8; 64]>()
+                    .await
+                {
+                    Ok(existing_secret) if existing_secret == secret => secret,
+                    Ok(_) => {
+                        return Err(anyhow::anyhow!(
+                            "Provided secret does not match existing secret"
+                        ))
+                        .map_err_cli_general()
+                    }
+                    Err(_) => {
+                        info!("Generating secret and writing to client storage");
```

Comment:
**suggestion:** As this repeats verbatim with lines 295-301, I would make a small free function for it `populate_with_random_secret` or something.

Reply:
It's not quite verbatim.. over here we have a provided secret to compare against.

Would you like a `populate_random_secret_if_missing(validation_secret: Option<[u8; 64]>)`? I personally don't like it.